RPP Glitch Tensor on HOST and HIP

HazarathKumarM opened the pull request:
Add glitch tensor implementation

r-abishek commented on src/modules/rppt_tensor_effects_augmentations.cpp:
```diff
@@ -521,23 +606,24 @@ RppStatus rppt_water_host(RppPtr_t srcPtr,
                           RpptRoiType roiType,
                           rppHandle_t rppHandle)
 {
+
```
Pls remove extra line added here

r-abishek commented on src/modules/rppt_tensor_effects_augmentations.cpp:
```diff
@@ -667,6 +753,7 @@ RppStatus rppt_ricap_host(RppPtr_t srcPtr,
     return RPP_SUCCESS;
 }
 
```
Pls remove extra line

r-abishek commented on src/modules/rppt_tensor_effects_augmentations.cpp:
```diff
@@ -1175,6 +1262,79 @@ RppStatus rppt_non_linear_blend_gpu(RppPtr_t srcPtr1,
+/******************** glitch ********************/
+
+RppStatus rppt_glitch_gpu(RppPtr_t srcPtr,
+                            RpptDescPtr srcDescPtr,
```
Align as per rpp std

r-abishek commented on src/modules/rppt_tensor_effects_augmentations.cpp:
```diff
@@ -1175,6 +1262,79 @@ RppStatus rppt_non_linear_blend_gpu(RppPtr_t srcPtr1,
+Rpp32u paramIndex = 0;
+copy_param_uint(xOffsetR, rpp::deref(rppHandle), paramIndex++);
+copy_param_uint(yOffsetR, rpp::deref(rppHandle), paramIndex++);
+copy_param_uint(xOffsetG, rpp::deref(rppHandle), paramIndex++);
```
Lot of copies here, @sampath1117 I think we have to consider taking one 2D tensor called rgbOffsets[batchSize]. And access as rgbOffsets[n].rOffset.x, rgbOffsets[n].rOffset.y
A single argument is passed rgbOffsets, and it could be in pinned memory removing these copies.

r-abishek commented on utilities/rpp-performancetests/HIP_NEW/Tensor_hip_pkd3.cpp:
```diff
@@ -1,25 +1,3 @@
-
-The above copyright notice and this permission notice shall be included in
-all copies or substantial portions of the Software.
-
```
Please remove changes to any files under the rpp-unittests or rpp-performance tests folders. These will not be updated henceforth. Just the test_suite folder must be updated to add relevant QA/unit/perf tests.

r-abishek commented on utilities/rpp-unittests/HIP_NEW/testAllScript.sh:
```diff
@@ -18,7 +18,7 @@ cwd=$(pwd)
 # Input Images - Three images (224 x 224)
 DEFAULT_SRC_FOLDER_1="$cwd/../TEST_IMAGES/three_images_224x224_src1"
-DEFAULT_SRC_FOLDER_2="$cwd/../TEST_IMAGES/three_images_224x224_src2"
+DEFAULT_SRC_FOLDER_2="$cwd/../TEST_IMAGES/three_images_224x224_src1"
```
Same comment here for all changes under the rpp-unittests / rpp-performancetests folders

r-abishek commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,1133 @@
+
+        if ((srcDescPtr->c == 3) && (srcDescPtr->layout == RpptLayout::NHWC) && (dstDescPtr->layout == RpptLayout::NCHW))
+        {
+            convert_pkd3_pln3<Rpp8u>(srcPtrImage, dstPtrImage, srcDescPtr, dstDescPtr, roi.xywhROI.roiWidth * 3);
```
Why aren't we using the internal shuffled loads and instead doing a preconvert here?

r-abishek commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,1133 @@
+                            {
+                                __m256i p;
+                                p = _mm256_loadu_si256((__m256i *)srcPtrTemp);
+                                _mm256_storeu_si256((__m256i *)dstPtrTemp, p);
```
Do we need 2 loads and 2 stores per vector? I think we should discuss this approach

r-abishek commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,319 @@
+{
+    if (roiType == RpptRoiType::LTRB)
+        hip_exec_roi_converison_ltrb_to_xywh(roiTensorPtrSrc, handle);
+    int localThreads_x = LOCAL_THREADS_X;
```
These 3 local threads lines should be removed and LOCAL_THREADS_X should be used directly. Pls check other functions

r-abishek commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,319 @@
+    int globalThreads_z = handle.GetBatchSize();
+    if ((srcDescPtr->layout == RpptLayout::NCHW) && (dstDescPtr->layout == RpptLayout::NCHW))
+    {
+        hipLaunchKernelGGL(glitch_pln_tensor,
```
Pls refer to other funcitons and similarly add name _hip as per new std

r-abishek commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,319 @@
+                                      RpptROIPtr roiTensorPtrSrc)
+{
+
+    int id_x = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
```
Why aren't these hip kernels vectorized? Glitch ideally writes elements in sequence, same as the reads.

r-abishek commented:
@HazarathKumarM Where are we on this? Pls resolve conflicts and mention which comments are done. @sampath1117 Do we need to talk on this?

HazarathKumarM commented:
> @HazarathKumarM Where are we on this? Pls resolve conflicts and mention which comments are done. @sampath1117 Do we need to talk on this?

I Need to resolve the conflicts in this branch and need to the clean up. I will complete it today

HazarathKumarM commented:
I have addressed all the review comments and updated the branch with the latest changes from the develop branch. Could you please review the latest changes.

r-abishek commented:
@HazarathKumarM I have updated my glitch branch to be in sync with master after today's merge. 
There are 144 files changed currently showing in this PR.
Could you do a git pull upstream master on yours too so only the files you touched in this PR are showing as changed?

HazarathKumarM commented:
> @HazarathKumarM I have updated my glitch branch to be in sync with master after today's merge. There are 144 files changed currently showing in this PR. Could you do a git pull upstream master on yours too so only the files you touched in this PR are showing as changed?
Done

r-abishek commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,2886 @@
+                    dstRowPtrTempR = dstRowPtr;
+                    dstRowPtrTempG = dstRowPtr + dstDescPtr->strides.cStride;
+                    dstRowPtrTempB = dstRowPtr + 2 * dstDescPtr->strides.cStride;
+                    if(yR < roi.xywhROI.roiHeight && xR < roi.xywhROI.roiWidth)
```
@HazarathKumarM Why do we need so many if statements between L160-L186? Even if we need it, there is so much repeated code @sampath1117 please check.

r-abishek commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,2886 @@
+                    dstRowPtrTempB = dstPtrImage + idx * dstDescPtr->strides.hStride + 2 * dstDescPtr->strides.cStride;
+                    for(int x = 0; x < roi.xywhROI.roiWidth; x++)
+                    {
+                        *dstRowPtrTempB = *srcRowPtrTempB;
```
Same comment here, there are hundreds of lines of close-enough repeated code

r-abishek reviewed (changes requested):
@sampath1117 Please do a round of review on this PR and discuss both HOST and HIP approaches with @HazarathKumarM 
Outputs might be correct, but we have to check the approach here.
There are CPU kernels with 100s of lines of close-enough if-statements / for loops that could possibly be modularized with more efficient use of helper functions, or removed if not needed.
We can discuss next week.

r-abishek commented on include/rppdefs.h:
```diff
@@ -407,6 +407,16 @@ typedef struct
+    RppiPoint r;
+    RppiPoint g;
+    RppiPoint b;
+}RpptChannelOffsets;
```
@HazarathKumarM Is this addressed?

HazarathKumarM replied:
Done

r-abishek commented on include/rppt_tensor_effects_augmentations.h:
```diff
@@ -424,4 +424,56 @@ RppStatus rppt_ricap_gpu(RppPtr_t srcPtr, RpptDescPtr srcDescPtr, RppPtr_t dstPt
+ * \param [in] xOffset value for R channel (1D tensor of size batchsize with xOffsetR[n] >= 0 for each image in batch)
+ * \param [in] yOffset value for R channel (1D tensor of size batchsize with yOffsetR[n] >= 0 for each image in batch)
+ * \param [in] xOffset value for G channel (1D tensor of size batchsize with xOffsetG[n] >= 0 for each image in batch)
+ * \param [in] yOffset value for G channel (1D tensor of size batchsize with yOffsetG[n] >= 0 for each image in batch)
```
@HazarathKumarM I don't think these are addressed

HazarathKumarM replied:
Done

r-abishek commented:
@HazarathKumarM or @sampath1117 Please comment something on all comments above if they have been addressed or cannot address, or if they were fixed with algorithm change.
I have added a reply-comment on some that are not yet changed that came up under files changed.

r-abishek commented:
Pls resolve merge conflicts against develop.

HazarathKumarM commented on utilities/test_suite/rpp_test_suite_common.h:
```diff
@@ -53,7 +53,7 @@ using namespace cv;
 
 #define CUTOFF 1
-#define DEBUG_MODE 0
+#define DEBUG_MODE 1
```
Done

HazarathKumarM commented on utilities/test_suite/HIP/Tensor_hip.cpp:
```diff
@@ -696,6 +696,30 @@ int main(int argc, char **argv)
+            {
+                testCaseName = "glitch";
+                RpptChannelOffsets *rgbOffsets;
+                CHECK(hipMalloc(&rgbOffsets, batchSize * sizeof(RpptChannelOffsets)));
```
Done

HazarathKumarM commented on src/modules/rppt_tensor_effects_augmentations.cpp:
```diff
@@ -1110,14 +1170,14 @@ RppStatus rppt_gaussian_noise_gpu(RppPtr_t srcPtr,
+                                     RppPtr_t srcPtr2,
+                                     RpptDescPtr srcDescPtr,
+                                     RppPtr_t dstPtr,
+                                     RpptDescPtr dstDescPtr,
```
Done

HazarathKumarM commented on src/modules/rppt_tensor_effects_augmentations.cpp:
```diff
@@ -527,19 +587,19 @@ RppStatus rppt_water_host(RppPtr_t srcPtr,
+                                    dstDescPtr,
+                                    amplitudeXTensor,
+                                    amplitudeYTensor,
+                                    frequencyXTensor,
```
Done

HazarathKumarM commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,2886 @@
+/*
+Copyright (c) 2019 - 2023 Advanced Micro Devices, Inc. All rights reserved.
```
Done

HazarathKumarM commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,2886 @@
+
+        Rpp32u xOffsetRchn = rgbOffsets[batchCount].r.x;
+        Rpp32u yOffsetRchn = rgbOffsets[batchCount].r.y;
+        Rpp32u xOffsetGchn = rgbOffsets[batchCount].g.x;
```
Done

HazarathKumarM commented on include/rppt_tensor_effects_augmentations.h:
```diff
@@ -424,4 +424,56 @@ RppStatus rppt_ricap_gpu(RppPtr_t srcPtr, RpptDescPtr srcDescPtr, RppPtr_t dstPt
+ * \param [in] dstDescPtr destination tensor descriptor (Restrictions - numDims = 4, offsetInBytes >= 0, dataType = U8/F16/F32/I8, layout = NCHW/NHWC, c = same as that of srcDescPtr)
+ * \param [in] xOffset value for R channel (1D tensor of size batchsize with xOffsetR[n] >= 0 for each image in batch)
+ * \param [in] yOffset value for R channel (1D tensor of size batchsize with yOffsetR[n] >= 0 for each image in batch)
+ * \param [in] xOffset value for G channel (1D tensor of size batchsize with xOffsetG[n] >= 0 for each image in batch)
```
Done

HazarathKumarM commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,2886 @@
+    for (int batchCount = 0; batchCount < dstDescPtr->n; batchCount++)
+    {
+        __m128i maskR = _mm_setr_epi8(0, 0x80, 0x80, 3, 0x80, 0x80, 6, 0x80, 0x80, 9, 0x80, 0x80, 12, 0x80, 0x80, 15);
+        __m128i maskGB = _mm_setr_epi8(0x80, 1, 2, 0x80, 4, 5, 0x80, 7, 8, 0x80, 10, 11, 0x80, 13, 14, 0x80);
```
Done

r-abishek commented:
@sampath1117 @HazarathKumarM ToT merge needed. Please put this in a reviewable state

r-abishek reviewed (changes requested):
@HazarathKumarM @sampath1117 Please address comments

HazarathKumarM commented on src/include/cpu/rpp_cpu_simd.hpp:
```diff
@@ -1021,6 +1021,72 @@ inline void rpp_load48_u8pkd3_to_f32pln3_avx(Rpp8u *srcPtr, __m256 *p)
+    __m128i px[3];
+
+    px[0] = _mm_loadu_si128((__m128i *)srcPtrR + srcLocs[0]);        /* load [R01|R02|R03|R04|R05|R06|R07|R08|R09|R10|R11|R12|R13|R14|R15|R16] */
+    px[1] = _mm_loadu_si128((__m128i *)srcPtrG + srcLocs[1]);    /* load [G01|G02|G03|G04|G05|G06|G07|G08|G09|G10|G11|G12|G13|G14|G15|G16] */
```
Done

HazarathKumarM commented on utilities/test_suite/HIP/Tensor_hip.cpp:
```diff
@@ -65,7 +65,7 @@ int main(int argc, char **argv)
     bool additionalParamCase = (testCase == 8 || testCase == 21 || testCase == 23|| testCase == 24 || testCase == 40 || testCase == 41 || testCase == 49 || testCase == 54);
     bool kernelSizeCase = (testCase == 40 || testCase == 41 || testCase == 49 || testCase == 54);
-    bool dualInputCase = (testCase == 2 || testCase == 30 || testCase == 33 || testCase == 61 || testCase == 63 || testCase == 65 || testCase == 68);
+    bool dualInputCase = (testCase == 2 || testCase == 30 || testCase == 61 || testCase == 63);
```
Done

HazarathKumarM commented on utilities/test_suite/HIP/Tensor_hip.cpp:
```diff
@@ -346,18 +346,6 @@ int main(int argc, char **argv)
-    RpptROI *cropRoi, *patchRoi;
-    if(testCase == 33)
-    {
-        CHECK(hipHostMalloc(&cropRoi, batchSize * sizeof(RpptROI)));
```
Done

HazarathKumarM commented on utilities/test_suite/HIP/Tensor_hip.cpp:
```diff
@@ -712,6 +700,32 @@ int main(int argc, char **argv)
+                {
+                    testCaseName = "glitch";
+                    RpptChannelOffsets *rgbOffsets;
+                    CHECK(hipMalloc(&rgbOffsets, batchSize * sizeof(RpptChannelOffsets)));
```
Done

HazarathKumarM commented on utilities/test_suite/HIP/runTests.py:
```diff
@@ -203,8 +203,8 @@ def rpp_test_suite_parser_and_validator():
-    parser.add_argument("--case_start", type = int, default = caseMin, help = "Testing start case # - Range must be in [" + str(caseMin) + ":" + str(caseMax) + "]")
-    parser.add_argument("--case_end", type = int, default = caseMax, help = "Testing end case # - Range must be in [" + str(caseMin) + ":" + str(caseMax) + "]")
+    parser.add_argument("--case_start", type = int, default = 0, help = "Testing range starting case # - (0:87)")
+    parser.add_argument("--case_end", type = int, default = 87, help = "Testing range ending case # - (0:87)")
```
Done

HazarathKumarM commented on utilities/test_suite/HIP/runTests.py:
```diff
@@ -222,8 +222,8 @@ def rpp_test_suite_parser_and_validator():
-    if ((args.case_start < caseMin or args.case_start > caseMax) or (args.case_end < caseMin or args.case_end > caseMax)):
-        print(f"Starting case# and Ending case# must be in the {caseMin}:{caseMax} range. Aborting!")
+    if ((args.case_start < 0 or args.case_start > 87) or (args.case_end < 0 or args.case_end > 87)):
+        print("Starting case# and Ending case# must be in the 0:87 range. Aborting!")
```
Done

HazarathKumarM commented on utilities/test_suite/HIP/runTests.py:
```diff
@@ -327,7 +327,7 @@ def rpp_test_suite_parser_and_validator():
 
 # List of cases supported
-supportedCaseList = ['0', '1', '2', '4', '8', '13', '20', '21', '23', '29', '30', '31', '33', '34', '36', '37', '38', '39', '45', '46', '54', '61', '63', '65', '68', '70', '80', '82', '83', '84', '85', '86', '87', '88', '89']
+supportedCaseList = ['0', '1', '2', '4', '8', '13', '20', '21', '23', '29', '30', '31', '33', '34', '35', '36', '37', '38', '39', '45', '46', '54', '61', '63', '65', '68', '70', '80', '82', '83', '84', '85', '86', '87', '88', '89']
```
Done

HazarathKumarM commented on utilities/test_suite/rpp_test_suite_common.h:
```diff
@@ -982,6 +983,8 @@ void compare_outputs_pkd_and_pln1(Rpp8u* output, Rpp8u* refOutput, RpptDescPtr d
                 if(diff <= CUTOFF)
                     matchedIdx++;
+                else
+                    std::cout<<"\n row i "<<i<<" col j "<<j<<" ref val "<<(int)*outRefVal<<" outval "<<(int)*outVal;
```
removed this line as it is not necessary

HazarathKumarM commented on utilities/test_suite/rpp_test_suite_common.h:
```diff
@@ -1017,6 +1020,8 @@ void compare_outputs_pln3(Rpp8u* output, Rpp8u* refOutput, RpptDescPtr dstDescPt
                     if(diff <= CUTOFF)
                         matchedIdx++;
+                    else
+                    std::cout<<"\n row i "<<i<<" col j "<<j<<" chn "<<c<<" ref val "<<(int)*outRefVal<<" outval "<<(int)*outVal;
```
removed

HazarathKumarM commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,294 @@
+
+template <typename T>
+RppStatus hip_exec_glitch_tensor(T *srcPtr,
+                                     RpptDescPtr srcDescPtr,
```
Done

HazarathKumarM commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,294 @@
+    d_float24 dst_f24;
+    d_float16 rlocSrc_f16, glocSrc_f16, blocSrc_f16;
+
+    compute_glitch_locs_hip(id_x, id_y, id_z, rgbOffsets, roiTensorPtrSrc, &rlocSrc_f16, &glocSrc_f16, &blocSrc_f16);
```
Done

HazarathKumarM commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,294 @@
+    increment_f8.f4[1] = make_float4(4.0f, 5.0f, 6.0f, 7.0f);
+
+    rLocsX_f8.f4[0] = static_cast<float4>(id_x + rgbOffsets[id_z].r.x) + increment_f8.f4[0];
+    rLocsX_f8.f4[1] = static_cast<float4>(id_x + rgbOffsets[id_z].r.x) + increment_f8.f4[1];
```
Done

HazarathKumarM commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,294 @@
+    rLocsY_f8.f4[0] = static_cast<float4>(id_y + rgbOffsets[id_z].r.y);
+    rLocsY_f8.f4[1] = static_cast<float4>(id_y + rgbOffsets[id_z].r.y);
+    check_locs(id_z, rLocsX_f8.f4[0], rLocsY_f8.f4[0], rgbOffsets[id_z].r.x, rgbOffsets[id_z].r.y, roiTensorPtrSrc);
+    check_locs(id_z, rLocsX_f8.f4[1], rLocsY_f8.f4[1], rgbOffsets[id_z].r.x, rgbOffsets[id_z].r.y, roiTensorPtrSrc);
```
Done

HazarathKumarM commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,294 @@
+    gLocsX_f8.f4[1] = static_cast<float4>(id_x + rgbOffsets[id_z].g.x) + increment_f8.f4[1];
+    gLocsY_f8.f4[0] = static_cast<float4>(id_y + rgbOffsets[id_z].g.y);
+    gLocsY_f8.f4[1] = static_cast<float4>(id_y + rgbOffsets[id_z].g.y);
+    check_locs(id_z, gLocsX_f8.f4[0], gLocsY_f8.f4[0], rgbOffsets[id_z].g.x, rgbOffsets[id_z].g.y, roiTensorPtrSrc);
```
Done

HazarathKumarM commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,736 @@
+#pragma omp parallel for num_threads(dstDescPtr->n)
+    for (int batchCount = 0; batchCount < dstDescPtr->n; batchCount++)
+    {
+        __m256i rMask = _mm256_setr_epi8(0, 0x80, 0x80, 3, 0x80, 0x80, 6, 0x80, 0x80, 9, 0x80, 0x80, 12, 0x80, 0x80, 15, 0x80, 0x80, 18, 0x80, 0x80, 21, 0x80, 0x80, 24, 0x80, 0x80, 27, 0x80, 0x80, 0x80, 0x80);
```
Done

HazarathKumarM commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,736 @@
+                    r = _mm256_loadu_epi8(srcPtrChannel + glitchSrcLocArray[0]);
+                    g = _mm256_loadu_epi8(srcPtrChannel + glitchSrcLocArray[1]);
+                    b = _mm256_loadu_epi8(srcPtrChannel + glitchSrcLocArray[2]);
+                    r = _mm256_shuffle_epi8(r, rMask);
```
Add rpp_simd_load routine

HazarathKumarM commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,736 @@
+                {
+                    compute_src_loc(dstLocRow, i, glitchSrcLocArray, srcDescPtr, rgbOffsets, roi, batchCount, 3);
+                    for (int c = 0; c < 3; c++)
+                    {
```
Done

HazarathKumarM commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,736 @@
+        {
+            Rpp8u *dstPtrRow;
+            dstPtrRow = dstPtrChannel;
+            Rpp32u alignedLength = ((int)((roi.xywhROI.roiWidth * 0.75)) / 8) * 8;   // Align dst width to process 16 dst pixels per iteration
```
Done

HazarathKumarM commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,294 @@
+    rlocSrc_f16->f4[0] = rLocsX_f8.f4[0];
+    rlocSrc_f16->f4[1] = rLocsX_f8.f4[1];
+    rlocSrc_f16->f4[2] = rLocsY_f8.f4[0];
+    rlocSrc_f16->f4[3] = rLocsY_f8.f4[1];
```
We have a function called "check_locs" where we verify whether individual locations are within the range or not. Currently, as we have implemented it with f4, there are 4 checks. If we change it to f8, we will need to include 8 condition checks, which will increase the code length

sampath1117 commented on include/rppt_tensor_effects_augmentations.h:
```diff
@@ -515,4 +515,46 @@ RppStatus rppt_gaussian_noise_voxel_gpu(RppPtr_t srcPtr, RpptGenericDescPtr srcD
+ */
+RppStatus rppt_glitch_gpu(RppPtr_t srcPtr, RpptDescPtr srcDescPtr, RppPtr_t dstPtr, RpptDescPtr dstDescPtr, RpptChannelOffsets *rgbOffsets, RpptROIPtr roiTensorPtrSrc, RpptRoiType roiType, rppHandle_t rppHandle);
+#endif // GPU_SUPPORT
+#endif // RPPT_TENSOR_EFFECTS_AUGMENTATIONS_H
```
Please add line at EOF

HazarathKumarM replied:
Done

sampath1117 commented on src/include/cpu/rpp_cpu_simd.hpp:
```diff
@@ -185,6 +185,15 @@ const __m256i avx_pxShufflePkd = _mm256_setr_m128(xmm_pxStore4Pkd, xmm_pxStore4P
+const __m256i avx_pxMaskG = _mm256_setr_epi8(0x80, 1, 0x80, 0x80, 4, 0x80, 0x80, 7, 0x80, 0x80, 10, 0x80, 0x80, 13, 0x80, 0x80, 16, 0x80, 0x80, 19, 0x80, 0x80, 22, 0x80, 0x80, 25, 0x80, 0x80, 28, 0x80, 0x80, 0x80);
+const __m256i avx_pxMaskB = _mm256_setr_epi8(0x80, 0x80, 2, 0x80, 0x80, 5, 0x80, 0x80, 8, 0x80, 0x80, 11, 0x80, 0x80, 14, 0x80, 0x80, 17, 0x80, 0x80, 20, 0x80, 0x80, 23, 0x80, 0x80, 26, 0x80, 0x80, 29, 0x80, 0x80);
+
+const __m256 avx_pxMaskR_ps = _mm256_setr_ps(-1.0f, 0, 0, -1.0f, 0, 0, 0, 0);
```
Are these masks used for shuffling or this is used for some compute?

HazarathKumarM replied:
I am using these masks for and operation. I am using this AND instruction to get the vector similar to shuffle instruction(_mm256_shuffle_epi8)

sampath1117 commented on src/include/cpu/rpp_cpu_simd.hpp:
```diff
@@ -1021,6 +1030,111 @@ inline void rpp_load48_u8pkd3_to_f32pln3_avx(Rpp8u *srcPtr, __m256 *p)
+
+    px[0] = _mm_loadu_si128((__m128i *)(srcPtr + srcLocs[1]));      /* load [R01|G01|B01|R02|G02|B02|R03|G03|B03|R04|G04|B04|R05|G05|B05|R06] - Need G01-04 */
+    px[1] = _mm_loadu_si128((__m128i *)(srcPtr + srcLocs[1] + 12)); /* load [R05|G05|B05|R06|G06|B06|R07|G07|B07|R08|G08|B08|R09|G09|B09|R10] - Need G05-08 */
+    p[1] = _mm256_cvtepi32_ps(_mm256_setr_m128i(_mm_shuffle_epi8(px[0], xmm_pxMaskG), _mm_shuffle_epi8(px[1], xmm_pxMaskG)));   /* Contains R01-08 */
```
Does p[1] at L1042 contain R01-R08?

HazarathKumarM replied:
Done

sampath1117 commented on src/include/cpu/rpp_cpu_simd.hpp:
```diff
@@ -1021,6 +1030,111 @@ inline void rpp_load48_u8pkd3_to_f32pln3_avx(Rpp8u *srcPtr, __m256 *p)
+
+    px[0] = _mm_loadu_si128((__m128i *)(srcPtr + srcLocs[2]));      /* load [R01|G01|B01|R02|G02|B02|R03|G03|B03|R04|G04|B04|R05|G05|B05|R06] - Need B01-04 */
+    px[1] = _mm_loadu_si128((__m128i *)(srcPtr + srcLocs[2] + 12)); /* load [R05|G05|B05|R06|G06|B06|R07|G07|B07|R08|G08|B08|R09|G09|B09|R10] - Need B05-08 */
+    p[2] = _mm256_cvtepi32_ps(_mm256_setr_m128i(_mm_shuffle_epi8(px[0], xmm_pxMaskB), _mm_shuffle_epi8(px[1], xmm_pxMaskB)));   /* Contains R01-08 */
```
Same comment
Does p[2] at L1046 contain R01-R08?

HazarathKumarM replied:
Done

sampath1117 commented on src/include/cpu/rpp_cpu_simd.hpp:
```diff
@@ -1021,6 +1030,111 @@ inline void rpp_load48_u8pkd3_to_f32pln3_avx(Rpp8u *srcPtr, __m256 *p)
+        p128[1] = _mm_loadu_ps(srcPtr + srcLocs[i] + 3);        /* load [R02|G02|B02|R03] - Need R01 */
+        p128[2] = _mm_loadu_ps(srcPtr + srcLocs[i] + 6);        /* load [R03|G03|B03|R04] - Need R03 */
+        p128[3] = _mm_loadu_ps(srcPtr + srcLocs[i] + 9);        /* load [R04|G04|B04|R05] - Need R04 */
+        _MM_TRANSPOSE4_PS(p128[0], p128[1], p128[2], p128[3]);  /* Contains R01-04 */
```
We just need only 1 value from each of 1054-1057? If yes, can you try with using a 1 setr instruction instead of using 4 loads followed by a transpose here?

HazarathKumarM replied:
Done

sampath1117 commented on src/include/cpu/rpp_cpu_simd.hpp:
```diff
@@ -1021,6 +1030,111 @@ inline void rpp_load48_u8pkd3_to_f32pln3_avx(Rpp8u *srcPtr, __m256 *p)
+    p[2] = _mm256_cvtepi32_ps(_mm256_setr_m128i(_mm_shuffle_epi8(px[0], xmm_pxMaskB), _mm_shuffle_epi8(px[1], xmm_pxMaskB)));   /* Contains R01-08 */
+}
+
+inline void rpp_glitch_load30_u8pkd3_to_u8pkd3_avx(Rpp8u *srcPtr, int * srcLocs, __m256i &p)
```
Change int * srcLocs to int *srcLocs

HazarathKumarM replied:
Done

sampath1117 commented on src/include/cpu/rpp_cpu_simd.hpp:
```diff
@@ -1021,6 +1030,111 @@ inline void rpp_load48_u8pkd3_to_f32pln3_avx(Rpp8u *srcPtr, __m256 *p)
+inline void rpp_glitch_load30_u8pkd3_to_u8pkd3_avx(Rpp8u *srcPtr, int * srcLocs, __m256i &p)
+{
+    __m256i px[3];
+    px[0] = _mm256_loadu_si256((__m256i *)(srcPtr + srcLocs[0]));
```
Please add inline comments on what is happening for these instructions

HazarathKumarM replied:
Done

sampath1117 commented on src/include/hip/rpp_hip_common.hpp:
```diff
@@ -2514,6 +2514,26 @@ __device__ __forceinline__ void rpp_hip_interpolate8_nearest_neighbor_pln1(T *sr
 }
 
+template <typename T>
+__device__ __forceinline__ void rpp_hip_load1_glitch(T *srcPtr, uint srcStrideH, float locSrcX, float locSrcY, int4 *roiPtrSrc_i4, float *dst, int channelStride, int chn)
```
if the function is only specific for glitch
suggest moving this inside the hip kernel file itself instead of adding in common header

HazarathKumarM replied:
Done

sampath1117 commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,299 @@
+        hip_exec_roi_converison_ltrb_to_xywh(roiTensorPtrSrc, handle);
+    int globalThreads_x = dstDescPtr->strides.hStride;
+    int globalThreads_y = dstDescPtr->h;
+    int globalThreads_z = handle.GetBatchSize();
```
Use the dstDescPtr->n for globalThreads_z instead of getting from handle since we are foreseeing some changes w.r.t handle in the future

sampath1117 commented on src/modules/rppt_tensor_effects_augmentations.cpp:
```diff
@@ -566,6 +566,66 @@ RppStatus rppt_non_linear_blend_host(RppPtr_t srcPtr1,
+    }
+    else if ((srcDescPtr->dataType == RpptDataType::F16) && (dstDescPtr->dataType == RpptDataType::F16))
+    {
+        glitch_f16_f16_host_tensor((Rpp16f*) (static_cast<Rpp8u*>(srcPtr) + srcDescPtr->offsetInBytes),
```
Please use reinterpret cast here instead of c style casting
check how it is done for new functions and make the change

HazarathKumarM replied:
Done

sampath1117 commented on src/modules/rppt_tensor_effects_augmentations.cpp:
```diff
@@ -1367,6 +1427,71 @@ RppStatus rppt_non_linear_blend_gpu(RppPtr_t srcPtr1,
+    }
+    else if ((srcDescPtr->dataType == RpptDataType::F16) && (dstDescPtr->dataType == RpptDataType::F16))
+    {
+        hip_exec_glitch_tensor((half*) (static_cast<Rpp8u*>(srcPtr) + srcDescPtr->offsetInBytes),
```
Same comment use reinterpret_cast<half *> here instead of c style (half *)

sampath1117 commented on src/modules/rppt_tensor_effects_augmentations.cpp:
```diff
@@ -1367,6 +1427,71 @@ RppStatus rppt_non_linear_blend_gpu(RppPtr_t srcPtr1,
+#endif // backend
+}
+
+/******************** water ********************/
```
why this definition added above water?
we generally add the new functions definition to the end. Please move this glitch definition also to the end

sampath1117 commented on utilities/test_suite/HIP/Tensor_hip.cpp:
```diff
@@ -712,6 +712,32 @@ int main(int argc, char **argv)
+                {
+                    testCaseName = "glitch";
+                    RpptChannelOffsets *rgbOffsets;
+                    CHECK(hipHostMalloc (&rgbOffsets, batchSize * sizeof(RpptChannelOffsets)));
```
Instead of using CHECK Macro
we are using CHECK_RETURN_STATUS macro in the recent changes for the validating the HIP API calls return values
Please use CHECK_RETURN_STATUS here

sampath1117 replied:
Would suggest to move this allocation and deallocation outside the loop
You need to check the test case outside followed by allocation and deallocation if test case is glitch

Something similar is done for here. Please refer
https://github.com/r-abishek/rpp/blob/daabb36f7ddbb8b2c5a501baacba2f1938e832ed/utilities/test_suite/HIP/Tensor_hip.cpp#L359
https://github.com/r-abishek/rpp/blob/daabb36f7ddbb8b2c5a501baacba2f1938e832ed/utilities/test_suite/HIP/Tensor_hip.cpp#L347

sampath1117 commented on utilities/test_suite/HIP/Tensor_hip.cpp:
```diff
@@ -712,6 +712,32 @@ int main(int argc, char **argv)
+                    else
+                        missingFuncFlag = 1;
+
+                    CHECK(hipHostFree(rgbOffsets));
```
Move this outside the loop as mentioned in previous comment

HazarathKumarM replied:
Done

sampath1117 commented on utilities/test_suite/HIP/Tensor_hip.cpp:
```diff
@@ -1303,4 +1329,4 @@ int main(int argc, char **argv)
     CHECK(hipFree(d_output));
     return 0;
-}
+}
```
Please revert this change

HazarathKumarM replied:
Done

sampath1117 commented on utilities/test_suite/HIP/runTests.py:
```diff
@@ -533,4 +533,4 @@ def rpp_test_suite_parser_and_validator():
     if checkFile:
         print("---------------------------------- Results of QA Test - Tensor_hip ----------------------------------\n")
-        print_qa_tests_summary(qaFilePath, supportedCaseList, nonQACaseList)
+        print_qa_tests_summary(qaFilePath, supportedCaseList, nonQACaseList)
```
Please revert this change

sampath1117 commented on utilities/test_suite/HOST/runTests.py:
```diff
@@ -494,4 +494,4 @@ def rpp_test_suite_parser_and_validator():
 
     for log_file in log_file_list:
-        print_performance_tests_summary(log_file, functionality_group_list, numRuns)
+        print_performance_tests_summary(log_file, functionality_group_list, numRuns)
```
Please revert this change

HazarathKumarM replied:
Done

sampath1117 commented on utilities/test_suite/rpp_test_suite_common.h:
```diff
@@ -849,7 +850,7 @@ inline void read_image_batch_turbojpeg(Rpp8u *input, RpptDescPtr descPtr, vector
         long jpegSize = ftell(fp);
         rewind(fp);
-        unsigned char* jpegBuf = (unsigned char*)malloc(jpegSize);
+        unsigned char* jpegBuf = (unsigned char*)calloc(jpegSize, sizeof(Rpp8u));
```
What is this change for?

HazarathKumarM replied:
We are having callocs every where in the test suite for initializing buffers except for this case and also It will store zeros by default to the buffer

sampath1117 commented on include/rppt_tensor_effects_augmentations.h:
```diff
@@ -515,4 +515,46 @@ RppStatus rppt_gaussian_noise_voxel_gpu(RppPtr_t srcPtr, RpptGenericDescPtr srcD
+ * \param [in] srcDescPtr source tensor descriptor (Restrictions - numDims = 4, offsetInBytes >= 0, dataType = U8/F16/F32/I8, layout = NCHW/NHWC, c = 1/3)
+ * \param [out] dstPtr destination tensor in HOST memory
+ * \param [in] dstDescPtr destination tensor descriptor (Restrictions - numDims = 4, offsetInBytes >= 0, dataType = U8/F16/F32/I8, layout = NCHW/NHWC, c = same as that of srcDescPtr)
+ * \param [in] rgbOffsets RGB offset values to use for the glitch augmentation (A single set of 3 Rppi point values that applies to all images in the batch)
```
Do we have a max limit for offset?
if yes, Can we mention them here?

HazarathKumarM replied:
Done

sampath1117 commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,673 @@
+
+    xB = col + rgbOffsets[batchCount].b.x;
+    yB = row + rgbOffsets[batchCount].b.y;
+
```
Would suggest to remove the additional blank lines at L32, L35, L38

sampath1117 replied:
is this addressed?
https://github.com/r-abishek/rpp/pull/199/files#r1593532428

sampath1117 commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,673 @@
+        {
+            Rpp8u *dstPtrRow;
+            dstPtrRow = dstPtrChannel;
+            Rpp32u alignedLength = ((int)((roi.xywhROI.roiWidth * 0.75)) / 8) * 8;   // Considering 75% of the ROI width to avoid loading out-of-bounds locations.
```
Please use static_cast instead of c style casting
Applicable to all such instances in this PR

HazarathKumarM replied:
Done

sampath1117 commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,673 @@
+                Rpp8u* dstRowPtrTempR = dstPtrRow;
+                Rpp8u* dstRowPtrTempG = dstPtrRow + dstDescPtr->strides.cStride;
+                Rpp8u* dstRowPtrTempB = dstPtrRow + 2 * dstDescPtr->strides.cStride;
+
```
Remove the empty blank line here at L104

HazarathKumarM replied:
Done

sampath1117 commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,673 @@
+            for (int dstLocRow = 0; dstLocRow < roi.xywhROI.roiHeight; dstLocRow++)
+            {
+                Rpp8u* dstPtrTemp = dstPtrRow;
+
```
Please remove blank lines like this

sampath1117 commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,673 @@
+                {
+                    __m256 p[6];
+                    compute_src_loc(dstLocRow, vectorLoopCount, glitchSrcLocArray, srcDescPtr, rgbOffsets, roi, batchCount, 1);
+                    rpp_simd_load(rpp_load48_u8pln3_to_f32pln3_avx, srcPtrChannel+glitchSrcLocArray[0], srcPtrChannel + srcDescPtr->strides.cStride + glitchSrcLocArray[1], srcPtrChannel + 2 * srcDescPtr->strides.cStride + glitchSrcLocArray[2], p);
```
Please add space before and after + operator
 srcPtrChannel + glitchSrcLocArray[0]

sampath1117 commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,673 @@
+                {
+                    __m256 p[6];
+                    compute_src_loc(dstLocRow, vectorLoopCount, glitchSrcLocArray, srcDescPtr, rgbOffsets, roi, batchCount, 1);
+                    rpp_simd_load(rpp_load48_u8pln3_to_f32pln3_avx, srcPtrChannel+glitchSrcLocArray[0], srcPtrChannel + srcDescPtr->strides.cStride + glitchSrcLocArray[1], srcPtrChannel + 2 * srcDescPtr->strides.cStride + glitchSrcLocArray[2], p);
```
if below are constant, would suggest precompute and store in a buffer prior and use it here instead of doing this operation everytime

```
srcDescPtr->strides.cStride + glitchSrcLocArray[1]
2 * srcDescPtr->strides.cStride + glitchSrcLocArray[2]
```

HazarathKumarM replied:
Done

sampath1117 commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,299 @@
+    rpp_hip_load8_glitch(srcPtr + srcIdx, srcStridesNH.y, &rlocSrc_f16, &srcRoi_i4, &(dst_f24.f8[0]), 3, 0);
+    rpp_hip_load8_glitch(srcPtr + srcIdx, srcStridesNH.y, &glocSrc_f16, &srcRoi_i4, &(dst_f24.f8[1]), 3, 1);
+    rpp_hip_load8_glitch(srcPtr + srcIdx, srcStridesNH.y, &blocSrc_f16, &srcRoi_i4, &(dst_f24.f8[2]), 3, 2);
+
```
Remove empty lines at L96 and L100

sampath1117 commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,299 @@
+    uint srcIdx = (id_z * srcStridesNH.x);
+    uint dstIdx = (id_z * dstStridesNH.x) + (id_y * dstStridesNH.y) + id_x * 3;
+
+    int4 srcRoi_i4 = *(reinterpret_cast<int4 *>(&roiTensorPtrSrc[id_z]));
```
Why do we need a reinterpret case here?
cannot we use them as uint in the code?

HazarathKumarM replied:
we are using int4 in all other HIP kernels also

sampath1117 reviewed (commented):
Please address review comments

r-abishek commented on docs/data/doxygenOutputs/geometric_augmentations_glitch_img150x150.png:
I have a question on this image. If are giving r/g/b offsets of 10/0/5 as per the test suite changes, both R and B are offset, and G remains the same.
This means I should not be able to see the correct image colors at all, but it seems that the original correct RGB image is in the background, with some R and G strokes overlayed. Is that right?

HazarathKumarM replied:
Yes Abishek
since the offsets are small in our case. We are able to see the correct image in the background.

sampath1117 commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,278 @@
+template <typename T>
+__device__ __forceinline__ void rpp_hip_load1_glitch(T *srcPtr, uint srcStrideH, float locSrcX, float locSrcY, int4 *roiPtrSrc_i4, float *dst, int channelStride, int chn)
+{
+    int srcIdx = locSrcY * srcStrideH + locSrcX * channelStride + chn;
```
where is roiPtrSrc_i4 used inside this function?
srcStrideH is uint and channelStride is int
can we combine these 2 into single uint2 variable?

HazarathKumarM replied:
Done

sampath1117 commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,278 @@
+template <typename T>
+__device__ __forceinline__ void rpp_hip_load1_glitch(T *srcPtr, uint srcStrideH, float locSrcX, float locSrcY, int4 *roiPtrSrc_i4, float *dst, int channelStride, int chn)
+{
+    int srcIdx = locSrcY * srcStrideH + locSrcX * channelStride + chn;
```
Please rename chn to channels

HazarathKumarM replied:
Done

sampath1117 commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,278 @@
+{
+    if (roiType == RpptRoiType::LTRB)
+        hip_exec_roi_converison_ltrb_to_xywh(roiTensorPtrSrc, handle);
+    int globalThreads_x = dstDescPtr->strides.hStride;
```
we need to have globalThread_x = (dstDescPtr->strides.hStride + 7) >> 3;
Since we are using 8 pixel load and store

Please change accordingly. Reference also added below
https://github.com/ROCm/rpp/blob/master/src/modules/hip/kernel/brightness.hpp#L186

HazarathKumarM replied:
Done

sampath1117 commented on utilities/test_suite/HIP/Tensor_hip.cpp:
```diff
@@ -369,6 +369,10 @@ int main(int argc, char **argv)
 
+    RpptChannelOffsets *rgbOffsets;
+    if(testCase == 35)
+        CHECK_RETURN_STATUS (hipHostMalloc (&rgbOffsets, batchSize * sizeof(RpptChannelOffsets)));
```
Please remove the additional space after CHECK_RETURN_STATUS, hipHostMalloc

HazarathKumarM replied:
Done

sampath1117 commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,648 @@
+{
+    RpptROI roiDefault = {0, 0, (Rpp32s)srcDescPtr->w, (Rpp32s)srcDescPtr->h};
+    omp_set_dynamic(0);
+#pragma omp parallel for num_threads(dstDescPtr->n)
```
@HazarathKumarM 
Please get the numThreads from handle instead using batchsize

Please ensure to follow this approach of getting numThreads from handle and use it for new CPU kernels henceforth
Reference added below
https://github.com/ROCm/rpp/blob/master/src/modules/cpu/kernel/brightness.hpp#L41

It needs to be done for all bit depth variants in HOST

HazarathKumarM replied:
Done

sampath1117 reviewed (commented):
Please address final review comments

HazarathKumarM commented on include/rppt_tensor_effects_augmentations.h:
```diff
@@ -515,4 +515,46 @@ RppStatus rppt_gaussian_noise_voxel_gpu(RppPtr_t srcPtr, RpptGenericDescPtr srcD
+ * \param [in] srcDescPtr source tensor descriptor (Restrictions - numDims = 4, offsetInBytes >= 0, dataType = U8/F16/F32/I8, layout = NCHW/NHWC, c = 1/3)
+ * \param [out] dstPtr destination tensor in HIP memory
+ * \param [in] dstDescPtr destination tensor descriptor (Restrictions - numDims = 4, offsetInBytes >= 0, dataType = U8/F16/F32/I8, layout = NCHW/NHWC, c = same as that of srcDescPtr)
+ * \param [in] rgbOffsets RGB offset values to use for the glitch augmentation (A 1D tensor in pinned/HOST memory contains single set of 3 Rppi point values that applies to all images in the batch. 0 < point.x <width  and 0 < point.y < height for each image in batch)
```
Done

HazarathKumarM commented on src/include/cpu/rpp_cpu_simd.hpp:
```diff
@@ -1021,6 +1025,96 @@ inline void rpp_load48_u8pkd3_to_f32pln3_avx(Rpp8u *srcPtr, __m256 *p)
+    {
+        Rpp32f *srcPtrTemp = srcPtr + srcLocs[i];
+        p[i] = _mm256_setr_ps(*srcPtrTemp, *(srcPtrTemp + 3), *(srcPtrTemp + 6), *(srcPtrTemp + 9), 
+                              *(srcPtrTemp + 12), *(srcPtrTemp + 15), *(srcPtrTemp + 18), *(srcPtrTemp + 21));
```
Done

r-abishek commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,278 @@
+__device__ void compute_glitch_locs_hip(int id_x, int id_y, RpptChannelOffsets rgbOffsets, RpptROI roiTensorPtrSrc, d_float24 *srcLocsX_f24, d_float24 *srcLocsY_f24)
+{
+    d_float8 increment_f8;
+    increment_f8.f4[0] = make_float4(0.0f, 1.0f, 2.0f, 3.0f);
```
increment_f8 is initialized, and only f4 is used? 
Pls just use float4 increment_f4;

HazarathKumarM replied:
Done

r-abishek commented on src/modules/hip/kernel/glitch.hpp:
```diff
@@ -0,0 +1,278 @@
+#include "rpp_hip_common.hpp"
+
+template <typename T>
+__device__ __forceinline__ void rpp_hip_load1_glitch(T *srcPtr, uint2 srcStrideCH, float locSrcX, float locSrcY, float *dst, int channels)
```
Ideally call by reference here as - float &locSrcX, float &locSrcY, float &dst

HazarathKumarM replied:
We don't need call by reference for locSrcX and locSrcY as we are not using those after this code. However, I am calling this function like this, where dst is passed by reference:
`rpp_hip_load1_glitch(srcPtr, srcStrideCH, srcX_f8->f1[0], srcY_f8->f1[0], &(dst_f8->f1[0]), channels);`

HazarathKumarM replied:
Done

r-abishek commented on src/modules/cpu/kernel/glitch.hpp:
```diff
@@ -0,0 +1,657 @@
+                Rpp8u* dstRowPtrTempR = dstPtrRow;
+                Rpp8u* dstRowPtrTempG = dstPtrRow + dstDescPtr->strides.cStride;
+                Rpp8u* dstRowPtrTempB = dstPtrRow + 2 * dstDescPtr->strides.cStride;
+                for (int vectorLoopCount = 0; vectorLoopCount < alignedLength; vectorLoopCount += 8)
```
Ideally the __AVX2__ flag should be added on all these vector loops

HazarathKumarM replied:
Done
